### ISSUE TITLE
Context Support

### ISSUE THREAD

**Issue by AncientHodler** [shopspring/decimal#206]
I wanted to ask if this library has something like a suport for a context.
A constext would be somethin like, setting a certain precision, certain round up for,..

Like i want all my math to be computed to the 250th decimal and no more. If i set the div precision to 250, doing multiple division would yield larger decimal number, and further doing math on those numbers would become increasingly time consuming.

My solution now would be to truncate all results to my give precision, or truncate the division function to 250 precision.

Or i could implement the context feature myself, but i dont know for now how to handle errors.... need to look that up.

The easiest way would be to truncate the math operations to a precision higher 1 or two points less than the division precision.

Im asking because the Decimal implementation in python has the context mechanic, and the Cockroach apd decimal library also has it .... ill probably going to look there to see how they done it

**Comment by mwoss** [comment]
Nope, right now this library only operates on global variables such as `DivisionPrecision` and `MarshalJSONWithoutQuotes`. I was thinking about adding support for context, but it has to wait for the 2.0 release as it would introduce a breaking change.
There are many things that should have done differently, but doing a breaking changes after 1.0 release is not something advisable.

**Comment by AncientHodler** [comment]
why would it make a breaking change ? anyway i was thinking on modifying the basic math functions to truncate to a certain decimal precision i want and need. The thing with the context precision in cockroachdb is that a 250 precision is the sum of the numbers before and after the decimal point. So if the number is 10 million which is 8 numbers big, your left with only "242" decimal precision....

Been trying yesterday for 9 hours straight to use the cockroach db, but their math syntax is somehow combersome, and couldnt make the sinus function work at all with it. as opposed to you library which implemented it correctly.
So im gonna stick with you library and truncate my math operations.
gonna have to write a function for logarithm though as there isnt any ....

**Comment by AncientHodler** [comment]
You might want to check this:
https://arxiv.org/pdf/1410.7176 titled
Efficient implementation of elementary functions in the medium-precision range
Elementary functions being
exp, sin, cos, log, atan

"We describe a new implementation of the elementary transcendental functions exp,
sin, cos, log and atan for variable precision up to approximately 4096 bits. Compared
to the MPFR library, we achieve a maximum speedup ranging from a factor 3 for cos to
30 for atan. Our implementation uses table-based argument reduction together with
rectangular splitting to evaluate Taylor series. We collect denominators to reduce the
number of divisions in the Taylor series, and avoid overhead by doing all multiprecision
arithmetic using the mpn layer of the GMP library. Our implementation provides
rigorous error bounds."
Im reading the pdf, and looking if i can implement this myself using your library math syntax.

If we have the elementary functions implemented like this, it would be a major, win, as wed have logarithm, power function, and all trigonometric function.

**Comment by mwoss** [comment]
Right now I can't think of a _clean_ way to introducing context support for a decimal lib without not changing any of the existing method signatures. I would have to introduce a new type with all math operations redefined, pass context var as a parameter to methods, or just add it as a decimal struct field.
It could be also stored as global var (something to what Python implementation of the decimal is doing), but I have concerns it won't be the best idea to move on with. I would have to rethink that.

**Comment by nicktobey** [comment]
Being able to control`MarshalJSONWithoutQuotes` via some kind of context would be incredibly useful. Only using global variables can cause huge problems in multi-threaded code, or if multiple libraries depend on this module.

**Comment by Rudde** [comment]
> Being able to control`MarshalJSONWithoutQuotes` via some kind of context would be incredibly useful. Only using global variables can cause huge problems in multi-threaded code, or if multiple libraries depend on this module.

How could this work in practice?

**Comment by rickb777** [comment]
I am against adding `Context` to the API because it soon gets unwieldy IMO.

The `DivisionPrecision` issue #307 has an easy workaround (use `DivRound` instead of `Div`).

An alternative way to deal with `MarshalJSONWithoutQuotes` could be simply to add a secondary decimal type that does this by default, along with an easy way to convert between `decimal.Decimal` and this new type. (This doesn't need to be in the API; any API user can do this for themselves.) Each use-case can then choose between the two decimal types according to JSON requirement for it.
